test/integration/src/test/java/io/pravega/test/integration/ReaderGroupNotificationTest.java

```diff
@@ -189,19 +213,130 @@ public void testSegmentNotifications() throws Exception {
+            notificationResults.add(notification);
+        };
+        SegmentNotifier segmentNotifier = (SegmentNotifier) readerGroup.getSegmentNotifier(executor);
+        segmentNotifier.registerListener(l1);
+
+        long start = System.currentTimeMillis();
+        log.info("Segments found as: {}", controller.getCurrentSegments(SCOPE, streamName).get());
+        CompletableFuture.runAsync(() -> {
+            while (System.currentTimeMillis() - start < Duration.ofMinutes(1).toMillis()) {
+                try {
+                    writer1.writeEvent(routingKeys.get(random.nextInt(3)), "data").get();
+                    Thread.sleep(0);
+                } catch (Throwable e) {
+                    log.error("test exception: ", e);
+                    throw new IllegalStateException(e);
+                }
+            }
+        }, threadPoolExecutor)
+                .exceptionally(e -> {
+                    log.error("Failure: ", e);
+                    throw new IllegalStateException(e);
+                });
+        try {
+            Thread.sleep(2000);
+            scale(controller, SCOPE, streamName, 1, 1, threadPoolExecutor);
+            Thread.sleep(2000);
+        } catch (InterruptedException e) {
+            log.error("Interrupted: ", e);
+        }
+        Retry.withExpBackoff(10, 10, 100, 10000)
+                .retryingOn(IllegalStateException.class)
+                .throwingOn(RuntimeException.class)
+                .runAsync(() -> controller.getCurrentSegments(SCOPE, streamName)
+                        .thenAccept(streamSegments -> {
+                            try {
+                                Thread.sleep(1000);
+                                EventStreamReader<String> reader1 = clientFactory.createReader("readerId", "reader", new JavaSerializer<>(),
+                                        ReaderConfig.builder().initialAllocationDelay(0).build());
+                                for (int i = 0; i < NUM_EVENTS; i++) {
+                                    EventRead<String> result =  reader1.readNextEvent(1000);
+                                    assertNotNull(result);
+                                    assertTrue(result.getEvent().equals("data"));
+                                }
+                                segmentNotifier.pollNow();
+                                SegmentNotification initialSegmentNotification = notificationResults.poll();
+                                log.info("notification={}", initialSegmentNotification);
+                                readerGroup.resetReaderGroup(readerGroupConfig);
+                                readerGroup.readerOffline("readerId", null);
+                                long segmentSize = streamSegments.getSegments().size();
+                                log.info("segment_size={}", segmentSize);
+                                if (segmentSize > 3) {
+                                    log.info("success");
```

Comment:
This implies as soon as the segment size grows beyond 3 the test would stop, right?

Reply:
Yes Sandeep.

Reply:
This pattern of ending the method with throwing an exception is counter-intuitive.
Use `AssertExtensions.assertGreaterThanOrEquals(...)` to validate.

Reply:
Yes Andrei